src/vtarget-out/inv-syn/inv_syn.cc

```diff
@@ -0,0 +1,633 @@
+        (|__AMC__wrapper| |__Sw'__|)
+        (not (|%w%_n rst|     |__Sw__|))  ; why not removed?
+        (not (|%w%_n rst|     |__Sw'__|)) ; why not removed?
+        (|%w%_h| |__Sw__|)
+        (|%w%_h| |__Sw'__|)
+    )
+    (INV2 |__Sw'__|)
+  )
+)
+
+
+; inv2 /\ ~ assert => fail
+(rule 
+  (=> 
+    (and
+      (INV2             |__Sw__|)
+      (|__AMC__wrapper| |__Sw__|)
+      (|%w%_h| |__Sw__|)
+      (not (|__ASST__|  |__Sw__|))
+    )
+    fail)
+)
+
+(query fail :print-certificate true)
+
+)***";
+
+// a local helper function to produce a list of and
+std::string JoinListWithAnd(const std::vector<std::string> & l) {
+  if (l.size() == 0)
+    return "true";
+  else if(l.size() == 1)
+    return l[0];
+  return "(and " + Join(l, " ")+")";
+}
+
+size_t skipArgs(const std::string & stmt, size_t start) {
+  // need to fit 
+  auto pos = start;
+  ILA_ASSERT(stmt.at(pos) == '(') << "Expecting '('";
+  unsigned level = 0;
+  do {
+    if (stmt.at(pos) == '(')
+      level ++;
+    else if (stmt.at(pos) == ')')
+      level --;
+    pos ++;
+  } while(level != 0);
+  return pos;
+}
+
+// a local helper function to return the type of a smt (define-fun ) stmt 
+bool is_returntype_bool(const std::string & stmt) {
+  std::string beginning("(define-fun |");
+  ILA_ASSERT (stmt.find(beginning) == 0) << "Not a `(define-fun |` stmt";
+  auto pos =  beginning.length()+1;
+  auto end_func_name_pos = stmt.find('|',pos);
+  auto start_arg_list = stmt.find('(', end_func_name_pos);
+  auto end_arg_list = skipArgs(stmt, start_arg_list);
+  auto Bool_pos = stmt.find("Bool", end_arg_list);
+  auto Bitvec_pos = stmt.find("(_ BitVec", end_arg_list);
+  if (Bool_pos < Bitvec_pos)
+    return true;
+  return false;
+}
+
+
+// extract 
+bool extractSigDefFromLine(
+  const std::string & mod_name,
+  const std::string & line,
+  std::set<std::string> & s, 
+  std::vector<std::pair<std::string,bool>> & v) {
+  // list of (name, is bool?)
+  bool found_match_state = false;
+  std::string search_target_wire_num = "(define-fun |" + mod_name + "#";
+  std::string search_target_n = "(define-fun |" + mod_name + "_n ";
+  if (line.find(search_target_wire_num) == 0) { // begins with it
+    auto mark = line.find("; \\");
+    if (mark != line.npos) {
+      auto signame = line.substr(mark+strlen("; \\"));
+      // it should not contain EOL
+      //signame = ReplaceAll(ReplaceAll(signame, "\n",""),"\r",""); 
+      if( IN(signame, s ) ) {
+        s.erase(signame); // remove it from the set
+
+        auto num_start = line.substr(search_target_wire_num.size());
+        auto end = num_start.find('|');
+        auto num = num_start.substr(0,end);
+        auto smt_expr = "|" + mod_name + "#" + num + "|";
+
+        v.push_back(std::make_pair(smt_expr,is_returntype_bool(line))); // |mod#n|
+
+        found_match_state = true;
+      }
+    } // found "; \\"
+  } // end of exists signal definition type 1
+  else if (line.find(search_target_n) == 0) { // begins with it
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Inefficient usage of string::find() in condition; string::starts_with() could be faster.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)